tests/test_salesforce_incremental_table_reset.py

```diff
@@ -0,0 +1,141 @@
+import datetime
+import dateutil.parser
+import pytz
+from datetime import datetime, timedelta
+
+from tap_tester import runner, menagerie, connections
+
+from base import SalesforceBaseTest
+
+
+class SalesforceIncrementalTableReset(SalesforceBaseTest):
+    @staticmethod
+    def name():
+        return "tap_tester_salesforce_incremental_table_reset"
+
+    @staticmethod
+    def expected_sync_streams():
+        return {
+            'Account',
+            'Contact',
+            'User',
+        }
+
+    @staticmethod
+    def convert_state_to_utc(date_str):
+        """
+        Convert a saved bookmark value of form '2020-08-25T13:17:36-07:00' to
+        a string formatted utc datetime,
+        in order to compare aginast json formatted datetime values
+        """
+        date_object = dateutil.parser.parse(date_str)
+        date_object_utc = date_object.astimezone(tz=pytz.UTC)
+        return datetime.datetime.strftime(date_object_utc, "%Y-%m-%dT%H:%M:%SZ")
+
+    @staticmethod
+    def parse_date(date_value):
+        """
+        Pass in string-formatted-datetime, parse the value, and return it as an unformatted datetime object.
+        """
+        date_formats = {
+            "%Y-%m-%dT%H:%M:%S.%fZ",
+            "%Y-%m-%dT%H:%M:%SZ",
+            "%Y-%m-%dT%H:%M:%S.%f+00:00",
+            "%Y-%m-%dT%H:%M:%S+00:00",
+            "%Y-%m-%d"
+        }
+        for date_format in date_formats:
+            try:
+                date_stripped = datetime.strptime(date_value, date_format)
+                return date_stripped
+            except ValueError:
+                continue
+
+        raise NotImplementedError("Tests do not account for dates of this format: {}".format(date_value))
+
+    def test_run(self):
+        self.salesforce_api = 'BULK'
+
+        replication_keys = self.expected_replication_keys()
+
+        # SYNC 1
+        conn_id = connections.ensure_connection(self)
+
+        # Run in check mode
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # Select only the expected streams tables
+        expected_streams = self.expected_sync_streams()
+        catalog_entries = [ce for ce in found_catalogs if ce['tap_stream_id'] in expected_streams]
+        self.select_all_streams_and_fields(conn_id, catalog_entries)
+        streams_replication_methods = {stream: self.INCREMENTAL
+                                       for stream in expected_streams}
+        self.set_replication_methods(conn_id, catalog_entries, streams_replication_methods)
+
+        # Run a sync job using orchestrator
+        first_sync_record_count = self.run_and_verify_sync(conn_id)
+        first_sync_records = runner.get_records_from_target_output()
+
+        first_sync_bookmarks = menagerie.get_state(conn_id)
+
+        # UPDATE STATE for Table Reset
+        new_states = {'bookmarks': dict()}
+        stream_to_current_state = {stream : bookmark.get(self.expected_replication_keys()[stream].pop())
```

Comment:
This seems like a lot of lines of code to remove one bookmark.  I've seen deep copy used in other tests so you may want to look into why that may be preferred.  What about something like this?
```suggestion
        new_states =  first_sync_bookmarks
        new_states['bookmarks'].pop('Users')
        menagerie.set_state(conn_id, new_states)
```

Reply:
Also, I think it would be more maintainable and readable if we set store the reset stream in a variable and set it.  Then everywhere where we are referencing the 'Users' stream we can instead reference the name of the variable.  Maybe call it`reset_stream`.  Then it will be obvious why this stream is being treated differently and if we ever want to change the stream we only need to change it in one place.

Reply:
also made a deep copy to have a clean code